runtime/compiler/ilgen/Walker.cpp

```diff
@@ -5609,8 +5595,10 @@ break
          ListIterator<TR_VMField> fieldIter(fieldsInfoByIndex->getFields());
          for (TR_VMField *field = fieldIter.getFirst(); field; field = fieldIter.getNext())
             {
-            if ((field->modifiers & J9AccFinal) && !isFinalFieldFromSuperClasses(methodClass, field->offset))
+            if ((field->modifiers & J9AccFinal) && methodClass == jitGetDeclaringClassOfROMField(comp()->j9VMThread(), methodClass, field->shape))
                {
+               if (comp()->getOption(TR_TraceILGen))
+                  traceMsg(comp(), "added fence due to field %s \n", field->name);
                push(callNode->getFirstChild());
```

Comment:
Is this going to be fixed separately?  i.e   callNode->getAllocation()  or something like that.

Reply:
There are multiple places where this needs to be changed. It should be a separate issue. @liqunl : Do you agree?

Reply:
The change to the API from get/setFirstChild to get/setAllocation should be a separate issue for conceptual integrity. I leave it to @AlenBadel and @liqunl to agree amongst themselves who will make that PR - it should be done soon to avoid forgetting about it.

Reply:
I made a change in StringPeepholes and it works in Alen's test. I'll check for all the places that need the change and open a PR ASAP.